magnum_cluster_api/hacks.py

```diff
@@ -0,0 +1,54 @@
+# Copyright (c) 2024 VEXXHOST, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+"""
+The purpose of this module is to provide a list of hacks and workarounds
+in place for modifications done to existing Cluster API resources to
+address issues in pre-existing clusters without changing their entire
+ClusterClass.
+"""
+
+import pykube
+
+from magnum_cluster_api import objects, utils
+
+CERTIFICATE_EXPIRY_DAYS_FIX_APPLIED = False
+
+
+def set_certificate_expiry_days(
+    api: pykube.HTTPClient,
+):
+    global CERTIFICATE_EXPIRY_DAYS_FIX_APPLIED
+    if not CERTIFICATE_EXPIRY_DAYS_FIX_APPLIED:
+        kcpts = objects.KubeadmControlPlaneTemplate.objects(
+            api, namespace="magnum-system"
+        ).all()
+        for kcpt in kcpts:
+            rollout_before = kcpt.obj["spec"]["template"]["spec"].get(
+                "rolloutBefore", {}
+            )
+            if "certificatesExpiryDays" not in rollout_before:
+                kcpt.obj["spec"]["template"]["spec"].setdefault("rolloutBefore", {})
+                kcpt.obj["spec"]["template"]["spec"]["rolloutBefore"][
+                    "certificatesExpiryDays"
+                ] = 21
+
+                # NOTE(mnaser): Since the KubeadmControlPlaneTemplate is immutable, we need to
+                #               delete the object and re-create it.
+                kcpt.delete()
+                del kcpt.obj["metadata"]["uid"]
```

Comment:
@okozachenko1203 I'm concerned about this since we're deleting a resource and creating it again and if we somehow fail to recreate after the delete, the cluster can probably go into a pretty odd state.

Is there a way to avoid the delete by any chance or figure out a way to make this "reliable"?  Also, I think maybe we need to introduce locking to this as well in case multiple delete at the same time, then a create conflict happens at the same time, and then all data is lost.

Reply:
I think we don't need to care about conflicts for both deletion and creation.
`object.delete()` will ignore `ObjectDoesNotExist` error and we do create via server-side apply.
https://codeberg.org/hjacobs/pykube-ng/src/commit/c1ac2c97587249e418d04d91f1297d4ffe3631c3/pykube/objects.py#L167-L180

Reply:
ok, i think we need lock because deletion and creation is done inside one function in a row to avoid the deletion of newly created kcpt by another replica.

Reply:
https://github.com/vexxhost/magnum-cluster-api/pull/363/commits/7d64eb0fb8626d51dc2dbebf8c72cc50ef479213